internal/parse/parser.go

```diff
@@ -1,57 +1,140 @@
 package parse
 
-import "strings"
+import (
+	"strings"
+)
 
-// Parser is used to parse the SQLAir DSL.
 type Parser struct {
-	// input is the DSL statement to be parted.
 	input string
-
-	// lastParsedPos is the character position of the last parsed part.
-	lastParsedPos int
-
-	// pos is the current character position of the parser.
-	pos int
+	pos   int
 }
 
-// NewParser returns a reference to a new parser.
 func NewParser() *Parser {
 	return &Parser{}
 }
 
-// init initializes the parser.
+// init resets the state of the parser and sets the input string.
 func (p *Parser) init(input string) {
 	p.input = input
-	p.lastParsedPos = 0
 	p.pos = 0
 }
 
-func (p *Parser) Parse(input string) (*ParsedExpr, error) {
-	p.init(input)
-	return nil, nil
+// A checkpoint struct for saving parser state to restore later. We only use
+// a checkpoint within an attempted parsing of an part, not at a higher level
+// since we don't keep track of the parts in the checkpoint.
+type checkpoint struct {
+	parser *Parser
+	pos    int
+}
+
+// save takes a snapshot of the state of the parser and returns a pointer to a
+// checkpoint that represents it.
+func (p *Parser) save() *checkpoint {
+	return &checkpoint{
+		parser: p,
+		pos:    p.pos,
+	}
+}
+
+// restore sets the internal state of the parser to the values stored in the
+// checkpoint.
+func (cp *checkpoint) restore() {
+	cp.parser.pos = cp.pos
 }
 
-// ParsedExpr represents a parsed expression.
-// It has a representation of the original SQL statement in terms of QueryParts
+// advance moves the parser's index forward by one element.
+func (p *Parser) advance() bool {
+	p.skipSpaces()
+	if p.pos == len(p.input) {
+		return false
+	}
+	p.pos++
```

Comment:
Although we're not seeing the rest of the logic here, this feels a bit suspect.. we can't advance the pointer without knowing what we're advancing over. Every parse or skip function moves the point forward towards the next character, so by the time something is parsed in theory we should be at the next position to be parsed. If we advance blindly from there, we ignored something. That's double troubling if we're both ignoring spaces and blindly moving the pointer forward. Also, the skipping of spaces probably needs accounting, in the sense that we need to know if there was indeed a space there serving as a separator from the previous token, otherwise we'll need to know how the underlying SQL parser behaves (e.g. is "foo()" the same as "foo ()?"). Although we don't have control over the entire syntax, we have some control over ours and can enforce proper spacing around our own tokens. But we can't do that if we just optionally skip and blindly move forward, etc.  Makes sense?

Reply:
You're certainly right in it feeling suspect, in my prototype it meant it was impossible to parse consecutive unseparated expressions (e.g. `&Person&Address`). I've in-placed the advance function so that in subsequent PRs the `p.pos++` will only be executed if nothing is parsed on in that iteration.

To keep track of the spaces we could instead:
Start and end the IO parser functions by calling skipSpaces and when we generate the SQL for them we put spaces in at each end. That way the Bypass expressions contain no spaces before/after an IO part but they end up in the generated SQL.

This approach will mostly require changes in future PRs and for now we only remove the calls to skipSpaces from the Parse function. Does this sound like a reasonable approach to you?